File: webtransport-py/room.py
from stream import get_uid_from_msg
from typing import List, Dict
from utils import getLogger
from models import Client, ReliableConnection, FastUnreliableConnection, PLAYER_TYPE_FAMILY
from comm.game_server.gs_manager import GameServer, GameServerMessages, _terminate_game_server

Log = getLogger(__name__)


class RoomState:
    MATCHING = 'MATCHING'
    PLAYING = 'PLAYING'

class Room:
    def __init__(self, name):
        self.name = name
        self.survivor_uids = {}
        self.family_uids = {}
        self.state = RoomState.MATCHING
        self.game_server: GameServer = None
        self.clients: List[Client] = []
        self.reliable_connection = ReliableConnection(self.clients)
        self.fast_unreliable_connection = FastUnreliableConnection(self.reliable_connection)

    def add_client(self, client: Client):
        self.clients.append(client)
        self.reliable_connection.clients = self.clients
        self.fast_unreliable_connection.clients = self.clients

    def remove_client(self, client: Client):
        self.clients.remove(client)
        self.reliable_connection.clients = self.clients
        self.fast_unreliable_connection.clients = self.clients

    async def to_game_client(self, msg):
        Log.debug(f'TO-CLIENT: {msg}')
        if GameServerMessages.GO in msg: # and GOD
            # await fast_unreliable_connection.send_message_others(msg, get_uid_from_msg(msg))
            await self.fast_unreliable_connection.send_message_all(msg)
        # elif GameServerMessages.GOD in msg:
        #     await reliable_connection.send_message_others(msg, get_uid_from_msg(msg))
        elif GameServerMessages.CONNECT_SELF in msg:
            uid = get_uid_from_msg(msg)
            for client in self.clients:
                if client.uid == uid:
                    Log.info(f'client get game server uid: {client}')
                    await self.reliable_connection.send_msg_to(client, msg)
                    break
        elif GameServerMessages.CONNECT_OTHER in msg:
            await self.reliable_connection.send_message_others(msg, get_uid_from_msg(msg))
        elif GameServerMessages.DISCONNECT in msg:
            await self.reliable_connection.send_message_all(msg)
        elif GameServerMessages.GAME_OVER in msg:
            await self.reliable_connection.send_message_all(msg)
            self.end_game()
        else:
            await self.fast_unreliable_connection.send_message_all(msg)

    def set_game_server(self, gs: GameServer):
        self.game_server = gs
        self.state = RoomState.PLAYING

    def end_game(self):
        _terminate_game_server(self.game_server.pid)
        # stop_game_server(self.game_server.pid, game_server_stopped_clear_communication)

    def players_len(self):
        return len(self.family_uids) + len(self.survivor_uids)

    def has_player(self, c_uid) -> bool:
        if c_uid in self.family_uids.keys():
            return True
        if c_uid in self.survivor_uids.keys():
            return True
        return False

    def remove_player(self, client: Client):
        c_uid = client.uid
        if c_uid in self.family_uids.keys():
            self.remove_client(client)
            del self.family_uids[c_uid]
        if c_uid in self.survivor_uids.keys():
            self.remove_client(client)
            del self.survivor_uids[c_uid]

    def add_player(self, type: 0 | 1, client: Client):
        client.type = type
        c_uid = client.uid
        self.remove_player(client)
        self.add_client(client)
        if type == PLAYER_TYPE_FAMILY:
            self.family_uids[c_uid] = {'ready' : False}
        else:
            self.survivor_uids[c_uid] = {'ready' : False}

    def ready_players(self):
        ready_players = 0
        for value in self.family_uids.values():
            if value['ready']:
                ready_players += 1
        for value in self.survivor_uids.values():
            if value['ready']:
                ready_players += 1
        return ready_players

    def can_start_game(self):
        half_players_count = self.players_len() / 2
        return self.ready_players() > half_players_count

    def player_ready(self, c_uid):
        if c_uid in self.family_uids.keys():
            if not self.family_uids[c_uid]['ready']:
                self.family_uids[c_uid] = {'ready' : True}
        if c_uid in self.survivor_uids.keys():
            if not self.survivor_uids[c_uid]['ready']:
                self.survivor_uids[c_uid] = {'ready' : True}

    def __str__(self):
        ready_players = 0
        res = f'{self.name}.family'
        for key, value in self.family_uids.items():
            res += f'.{key}'
            if value['ready']:
                ready_players += 1

        res += '.survivors'
        for key, value in self.survivor_uids.items():
            res += f'.{key}'
            if value['ready']:
                ready_players += 1

        return res + f'.ready.{ready_players}'

File: game/src/main/java/ovh/look/game/models/Room.java
package ovh.look.game.models;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.logging.Logger;

import lombok.Data;

@Data
public class Room {
    private static final Logger Log = Logger.getLogger(Room.class.getName());

    // public static class RoomState {
    //     public static final String MATCHING = "MATCHING";
    //     public static final String PLAYING = "PLAYING";
    // }

    private String name;
    private Map<Integer, Map<String, Boolean>> survivorUids;
    private Map<Integer, Map<String, Boolean>> familyUids;
    private RoomState state;
    private GameServer gameServer;
    private List<Client> clients;
    private ReliableConnection reliableConnection;
    private FastUnreliableConnection fastUnreliableConnection;

    public Room(String name) {
        this.name = name;
        this.survivorUids = new HashMap<>();
        this.familyUids = new HashMap<>();
        this.state = RoomState.MATCHING;
        this.gameServer = null;
        this.clients = new ArrayList<>();
        this.reliableConnection = new ReliableConnection(this.clients);
        this.fastUnreliableConnection = new FastUnreliableConnection(this.reliableConnection);
    }

    public void addClient(Client client) {
        this.clients.add(client);
        this.reliableConnection.setClients(this.clients);
        this.fastUnreliableConnection.setClients(this.clients);
    }

    public void removeClient(Client client) {
        this.clients.remove(client);
        this.reliableConnection.setClients(this.clients);
        this.fastUnreliableConnection.setClients(this.clients);
    }

    public void toGameClient(String msg) {
        Log.info("TO-CLIENT: " + msg);

        try {
            if (msg.contains(GameServerMessages.GOD.getValue())) {
                this.reliableConnection.sendMessageAll(msg);
            } else if (msg.contains(GameServerMessages.GO.getValue())) {
                this.fastUnreliableConnection.sendMessageAll(msg);
            } else if (msg.contains(GameServerMessages.CONNECT_SELF.getValue())) {
                int uid = getUidFromMsg(msg);
                for (Client client : this.clients) {
                    if (client.getUid() == uid) {
                        Log.info("client get game server uid: " + client);
                        this.reliableConnection.sendMsgTo(client, msg);
                        break;
                    }
                }
            } else if (msg.contains(GameServerMessages.CONNECT_OTHER.getValue())) {
                this.reliableConnection.sendMessageOthers(msg, getUidFromMsg(msg));
            } else if (msg.contains(GameServerMessages.DISCONNECT.getValue())) {
                this.reliableConnection.sendMessageAll(msg);
            } else if (msg.contains(GameServerMessages.GAME_OVER.getValue())) {
                this.reliableConnection.sendMessageAll(msg);
                this.endGame();
            } else {
                this.reliableConnection.sendMessageAll(msg);
            }
        } catch (Exception e) {
            Log.severe("Error in toGameClient: " + e.getMessage());
        }
    }

    public void setGameServer(GameServer gs) {
        this.gameServer = gs;
        this.state = RoomState.PLAYING;
    }

    public void endGame() {
        for (Integer cUid : this.familyUids.keySet()) {
            this.familyUids.get(cUid).put("ready", false);
        }
        for (Integer cUid : this.survivorUids.keySet()) {
            this.survivorUids.get(cUid).put("ready", false);
        }
    }

    public int playersLen() {
        return this.familyUids.size() + this.survivorUids.size();
    }

    public boolean hasPlayer(int cUid) {
        return this.familyUids.containsKey(cUid) || this.survivorUids.containsKey(cUid);
    }

    public void removePlayer(Client client) {
        int cUid = client.getUid();
        if (this.familyUids.containsKey(cUid)) {
            this.removeClient(client);
            this.familyUids.remove(cUid);
        }
        if (this.survivorUids.containsKey(cUid)) {
            this.removeClient(client);
            this.survivorUids.remove(cUid);
        }
    }

    public void addPlayer(PlayerType type, Client client) {
        client.setType(type);
        int cUid = client.getUid();
        this.removePlayer(client);
        this.addClient(client);
        if (PlayerType.FAMILY.equals(type)) {
            this.familyUids.put(cUid, new HashMap<String, Boolean>() {{
                put("ready", false);
            }});
        } else {
            this.survivorUids.put(cUid, new HashMap<String, Boolean>() {{
                put("ready", false);
            }});
        }
    }

    public int readyPlayers() {
        int readyPlayers = 0;
        for (Map<String, Boolean> value : this.familyUids.values()) {
            if (value.get("ready")) {
                readyPlayers += 1;
            }
        }
        for (Map<String, Boolean> value : this.survivorUids.values()) {
            if (value.get("ready")) {
                readyPlayers += 1;
            }
        }
        return readyPlayers;
    }

    public boolean canStartGame() {
        double halfPlayersCount = this.playersLen() / 2.0;
        return this.readyPlayers() > halfPlayersCount;
    }

    public void playerReady(int cUid) {
        if (this.familyUids.containsKey(cUid)) {
            if (!this.familyUids.get(cUid).get("ready")) {
                this.familyUids.get(cUid).put("ready", true);
            }
        }
        if (this.survivorUids.containsKey(cUid)) {
            if (!this.survivorUids.get(cUid).get("ready")) {
                this.survivorUids.get(cUid).put("ready", true);
            }
        }
    }

    @Override
    public String toString() {
        int readyPlayers = 0;
        StringBuilder res = new StringBuilder(this.name + ".family");
        for (Map.Entry<Integer, Map<String, Boolean>> entry : this.familyUids.entrySet()) {
            int cUid = entry.getKey();
            boolean ready = entry.getValue().get("ready");
            res.append(".").append(cUid).append(".").append(ready ? 1 : 0);
            if (ready) {
                readyPlayers += 1;
            }
        }

        res.append(".survivors");
        for (Map.Entry<Integer, Map<String, Boolean>> entry : this.survivorUids.entrySet()) {
            int cUid = entry.getKey();
            boolean ready = entry.getValue().get("ready");
            res.append(".").append(cUid).append(".").append(ready ? 1 : 0);
            if (ready) {
                readyPlayers += 1;
            }
        }

        return res.append(".ready.").append(readyPlayers).toString();
    }

    // Placeholder methods for missing implementations
    private int getUidFromMsg(String msg) {
        return Integer.parseInt(msg.substring(0, msg.indexOf('.')));
    }
}
